lib/guest_configuration_base.pm

```diff
@@ -0,0 +1,1408 @@
+        diag("Common environment preparation is done now.");
+    }
+    else {
+        diag("Common environment preparation had already been done.");
+    }
+    return $self;
+}
+
+#Remove all existing guests and storage files in /var/lib/libvirt/images/
+sub clean_up_all_guests {
+    my $self = shift;
+
+    my @_guests_to_clean_up = split(/\n/, script_output("virsh list --all --name", proceed_on_failure => 1));
+    if (scalar(@_guests_to_clean_up) gt 0) {
+        diag("Going to clean up all guests on $self->{host_name}");
+        foreach (@_guests_to_clean_up) {
+            script_run("virsh destroy $_");
+            script_run("virsh undefine $_ --nvram") if (script_run("virsh undefine $_") ne 0);
+        }
+        script_run("rm -f -r /var/lib/libvirt/images/*");
+    }
+    else {
+        diag("No guests reside on this host $self->{host_name}");
+    }
+    return $self;
+}
+
+#Create individual guest log folder using its name
+sub prepare_guest_log_folder {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->{guest_log_folder} = $common_log_folder . '/' . $self->{guest_name};
+    script_run("rm -f -r $self->{guest_log_folder}");
+    assert_script_run("mkdir -p $self->{guest_log_folder}");
+    return $self;
+}
+
+#Configure [guest_domain_name] and [guest_name_options].User can still change [guest_name] and [guest_domain_name] by passing non-empty arguments using hash.
+sub config_guest_name {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->config_guest_params(@_) if (scalar(@_) gt 0);
+    if ($self->{guest_installation_result} eq '') {
+        $self->{guest_domain_name}  = 'testvirt.net' if ($self->{guest_domain_name} eq '');
+        $self->{guest_name_options} = "--name $self->{guest_name}";
+    }
+    return $self;
+}
+
+#Configure [guest_metadata_options].User can still change [guest_metadata] by passing non-empty arguments using hash.
+#If installation already passes,modify_guest_params will be called to modify [guest_metadata] using already modified [guest_metadata_options].
+sub config_guest_metadata {
+    my $self = shift;
+
+    $self->reveal_myself;
+    my $_guest_name = $self->{guest_name};
+    $self->config_guest_params(@_) if (scalar(@_) gt 0);
+    if ($self->{guest_metadata} ne '') {
+        $self->{guest_metadata_options} = "--metadata $self->{guest_metadata}";
+        $self->modify_guest_params($_guest_name, 'guest_metadata_options') if ($self->{guest_installation_result} eq 'PASSED');
+    }
+    return $self;
+}
+
+#Configure [guest_vcpus_options].User can still change [guest_vcpus] by passing non-empty arguments using hash.
+#If installations already passes,modify_guest_params will be called to modify [guest_vcpus] using already modified [guest_vcpus_options].
+sub config_guest_vcpus {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->config_guest_params(@_) if (scalar(@_) gt 0);
+    if ($self->{guest_vcpus} ne '') {
```

Comment:
What about giving some default value if not set by caller, to let users only needing to customize params that they want to set? Similar comment for guest memory and others wherever we can provide this convenience for users.